DigitalTwins: Switch to etags, drop tiny Options classes

tg-msft opened the pull request.

tg-msft committed:
Switch to etags, drop tiny Options classes

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs:
```diff
@@ -86,7 +88,26 @@ public class BasicDigitalTwin
         /// </summary>
         [JsonPropertyName(DigitalTwinsJsonPropertyNames.DigitalTwinETag)]
-        public string ETag { get; set; }
+        [JsonConverter(typeof(OptionalETagConverter))] // TODO: Remove when #16272 is fixed
```
this is interesting

tg-msft replied:
(And let's reference #16272 here so it shows up on the issue too)

azabbasi commented on sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs:
```diff
@@ -47,26 +47,18 @@ public async Task PublishTelemetry_Lifecycle()
-                    TimeStamp = default
-                };
-                Response publishTelemetryResponse = await client.PublishTelemetryAsync(roomWithWifiTwinId, Recording.Random.NewGuid().ToString(), "{\"Telemetry1\": 5}", telemetryOptions).ConfigureAwait(false);
+                Response publishTelemetryResponse = await client.PublishTelemetryAsync(roomWithWifiTwinId, Recording.Random.NewGuid().ToString(), "{\"Telemetry1\": 5}", timeStamp: default(DateTimeOffset)).ConfigureAwait(false);
```
nit: line seems too long. we can break it into multiple

tg-msft replied:
Split across lines

azabbasi commented on sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs:
```diff
@@ -47,26 +47,18 @@ public async Task PublishTelemetry_Lifecycle()
                 };
                 Response publishComponentTelemetryResponse = await client
-                    .PublishComponentTelemetryAsync(roomWithWifiTwinId, wifiComponentName, Recording.Random.NewGuid().ToString(), JsonSerializer.Serialize(telemetryPayload), componentTelemetryOptions)
+                    .PublishComponentTelemetryAsync(roomWithWifiTwinId, wifiComponentName, Recording.Random.NewGuid().ToString(), JsonSerializer.Serialize(telemetryPayload), timeStamp: default(DateTimeOffset))
```
same here.

tg-msft replied:
Split across lines

tg-msft committed:
PR Feedback

tg-msft commented:
It's also worth pointing out that we can switch from ["<= 6 parameters" simple params to an Options type complex param](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-parameters) in the future without breaking if the service adds many more parameters.

Assuming we had something today like:
```C#
public void Foo(string id, bool include = false) { ... }
```

We can switch to an Options bag in the future with:
```C#
[EditorBrowsable(Never)]
public void Foo(string id, bool include) { ... } // note: default value removed

public void Foo(string id, FooOptions options = default) { ... }

public class FooOptions
{
    public bool Include { get; set; }
    public string Random { get; set; }
    // + more new params...
}
```

And it would not be a breaking change.  Customers will continue to compile, upgrading without recompile is fine, etc.  We'd only see the new signature in intellisense and documentation so new customers would always start there.  The only pain is customers wanting to use the new `Random` property would have to switch from `Foo(id, include: true)` to `Foo(id, new FooOptions { Include = true, Random = "..." })`.  We consider that acceptable and have done it often in other services like Storage.

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:
```diff
@@ -1931,7 +2017,7 @@ public virtual Response DeleteEventRoute(string eventRouteId, CancellationToken
             string messageId,
             string payload,
-            PublishTelemetryOptions options = null,
+            DateTimeOffset? timeStamp = null,
```
I wonder if this should be cased `timestamp`. The only other case of timestamp I found in .NET is `Stopwatch.GetTimestamp`.

https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch.gettimestamp?view=netcore-3.1

tg-msft replied:
Krzysztof wrote `Stopwatch`.  I'll probably get yelled at if we don't follow it.  :smile:

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:
```diff
@@ -1940,6 +2026,12 @@ public virtual Response DeleteEventRoute(string eventRouteId, CancellationToken
                 messageId = Guid.NewGuid().ToString();
             }
 
+            PublishTelemetryOptions options = new PublishTelemetryOptions();
```
var

tg-msft replied:
Fixed.

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs:
```diff
@@ -86,7 +88,26 @@ public class BasicDigitalTwin
+            public override ETag? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                string value = reader.GetString();
+                return value != null ?
```
```csharp
return value != null
    ? new ETag(value)
    : (ETag?)null;
```

tg-msft replied:
Fixed.

tg-msft committed:
PR feedback

tg-msft committed:
Exporting APIs too

timtay-microsoft committed:
Add etag property to BasicRelationship, fix spacing in comments

Also factor out OptionalETagConverter since both BasicRelationship and BasicDigitalTwin use it

timtay-microsoft committed:
Add test that exercise ETags in create/update/delete APIs

One ignored test in the 412 case when updating components because the service doesn't throw a 412 exception when updating a component if the etag is out of date which is probably a service bug

timtay-microsoft committed:
fixup

timtay-microsoft committed:
fixup

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Serialization/OptionalETagConverter.cs:
```diff
@@ -0,0 +1,31 @@
+        public override void Write(Utf8JsonWriter writer, ETag? value, JsonSerializerOptions options)
+        {
+            if (value == null)
+            { writer.WriteNullValue(); }
```
won't this NRE?

tg-msft replied:
If the `writer` is null?  That won't happen with STJ and it's not a publicly exposed type.  We can add an `Argument.NotNull(writer, nameof(writer));` if we think it's possible customers could somehow get here.

drwill-ms replied:
Ah, I misread this as `if (write == null)` for some reason.

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/tests/ComponentTests.cs:
```diff
@@ -93,5 +93,191 @@ public async Task Component_Lifecycle()
+
+                // Patch component again, but with the now out of date ETag
+                JsonPatchDocument secondComponentUpdatePatchDocument = new JsonPatchDocument();
+                componentUpdatePatchDocument.AppendReplace("/Network", "Even newer Network");
```
Setting the wrong JsonPatchDocument here
